### ISSUE TITLE
Release planning

### ISSUE THREAD

**Issue by tbekolay** [nengo/nengo#1114]
Currently, we don't have a process for deciding what goes in the next release. In the past, we've basically added every new issue and PR to the next release, which contributed to the year-long gap between 2.0 and 2.1. We want to have a much quicker release cycle, so we need a way to decide what goes in each release.

Please add your proposal as a comment to this issue if you think something else would be better. We'll talk about it and vote at the next dev meeting!

**Comment by tbekolay** [comment]
Here's what I propose:

1. New PRs are not assigned to a milestone.
2. Issues are never assigned to a milestone.
3. Only existing PRs can be considered for feature releases.
4. Once a release happens, the maintainers set a time to meet and plan the next release. (question: should this meeting be at a normal dev meeting, or a separate meeting? If separate, maintainers have to be there, and others can attend if interested.)

The release planning meeting goes as follows:

- If there are bug fix PRs, then the next release should be a bugfix release (i.e., increase the patch version).
  - Before the meeting, I'll add any bugfix PRs to the next release milestone. Maintainers and other interested parties look through the current open issues to see if there are any unfixed bugs that we should really address for the bugfix release.
  - At the meeting, we'll go over the PRs in the milestone and discuss any additional unfixed bugs that we should fix for the release.
- If there are no bug fix PRs, then the next release should be a feature release (increases the minor version).
  - Before the meeting, maintainers and other interested parties choose the PRs they think should be in the next release. Features that don't have an associated PR yet won't be considered.
  - At the meeting, each person will pitch the PRs they think should be considered. Consensus will decide what ends up going in.

This plan is skewed toward keeping the issue / PR queue small, which is my main goal at the moment.

**Comment by drasmuss** [comment]
It's a bit early to say whether we could achieve this, but I'm somewhat hopeful that with a few new ABR reviewers, and our new "revise and resubmit" policies for clearing out old PRs, we will no longer have this pool of unresolved PRs that we need to organize into releases.  

In other words, I think in an ideal world when we come to step 3 above there aren't going to be a bunch of existing PRs to be considered for future releases.  They'll just be steadily moving into master as they are completed.

So in that scenario, I think releases would be more of a retroactive thing.  At e.g. each dev meeting (or every other dev meeting), we could look at how far master has diverged from the latest release.  Then we can make a call about whether it's worthwhile to push out a new release (and if it should be a minor or patch increment).

Motivations for this approach:

1. I think past experience has shown us that it is difficult to anticipate which changes we will want in a release.  Features that we thought would take a long time get implemented quickly, or vice versa.  New bugs crop up and get fixed.  People have more or less time to work on a WIP PR.

2. Also in the past I think we have seen releases slowed down a lot because they were 90% complete, but we were waiting for one or two PRs on the release list.  I think it'd be better to just make a release with that 90%, and let the 10% make it into a release whenever they get done (in which case we're just doing the retroactive release thing in practice anyway).

**Comment by drasmuss** [comment]
Note that in this scenario we would still require some planning/organization system to decide which features we'd like implemented, and which bugs we need fixed.  I'm just suggesting that that be decoupled from the release schedule.

**Comment by hunse** [comment]
Personally, I think we should have two separate systems. One system should track how critical outstanding bugs/issues are. It would not assign these to any release, but would basically be a way to prioritize what we're working on. Like with @tbekolay's proposal, issues would never be assigned to a release, just given a priority.

I think I mostly agree with @tbekolay's proposal with respect to putting PRs into releases. The main place where I differ is that I'm not sure if "maintainers and other interested parties [should] look through the open issues" to see if there's anything we want to put in the release. That's too close to assigning issues to releases for me, and I'm worried that we'll end up holding up releases because someone thinks some bug should be fixed for it. I think this is what @tbekolay is suggesting for feature PRs, I'm just saying the same should apply to bugfix PRs.

One other thing I think we should consider is how different feature additions will affect backends. I'm not quite sure how to categorize this, but some rough categories would be:
1. Features that break all backends.
2. Features that break reference builder backends.
3. Features that break all backends for models that use the new feature.
4. Features that break reference builder backends for models that use the new feature.
5. Features that break no backends.

It might be the case that we need to merge some PRs, particularly those falling into the first few of these categories, into branches other than master, that we then release later. This would allow us to keep merging bug fixes into master and doing less serious releases, while holding off serious releases so they don't happen too often (since they are more work for backend maintainers).

The other option would be to leave such PRs unmerged, but I don't like that since then it's a) harder for other people to use them in synchrony with other pending features, and b) it's harder to develop off of those pending features. We can also make sure we keep the pending feature branch up to date with master.

**Comment by jgosmann** [comment]
* I think there might be certain types of critical bugs/issues that should be fixed in a timely manner. So we should have some way to keep track of the priority of bugs.
* Some bugs/issues might be that critical that they should hold up a release and I think these should be in the corresponding milestone (also to make sure that the GitHub progress bar only gets to 100% when all important issues ares resolved).
* Depending on the exact process, I'm fine with leaving out less important issues from milestones.

> It might be the case that we need to merge some PRs, particularly those falling into the first few of these categories, into branches other than master, that we then release later.

I might be in favor of having someting like [git-flow], maybe a bit simpler, but with a develop and a master branch so that we can merge new features and do some integration testing while still having the possibility to do bugfix releases.

**Comment by hunse** [comment]
> Some bugs/issues might be that critical that they should hold up a release

The only bugs that would fall into this category for me are a) regressions, and b) bugs in things in master that have not been released yet. Anything else to me should have a priority, but if it's high priority and hasn't been fixed yet, and we want to do a release, then I don't see why we wouldn't just do a release and then do another one e.g. a week later when the critical bug has been fixed.

This is a bit different for features. If we're doing a feature release, I can see saying that we want to hold up a release while we wait for a feature to get in, because it might be longer before we get another feature release out (especially if both the feature in question and the other features in the release break backends). However, the thing I worry about is people really wanting something implemented and so "nominating" it for the next release. If you really want something fixed, fix it. Then once there's a PR, that PR can be put under the next release milestone.

**Comment by tbekolay** [comment]
Yes, re: @drasmuss and @hunse, I was going to say that the main issue with merging PRs into master as they come in is that some PRs affect backends and some things don't, and it's not clear the best way to deal with these things release wise. I want to avoid the case where we force someone to upgrade in order to fix a bug they're experiencing if that upgrade also means they can't use SpiNNaker or OCL or the GUI anymore. That was my also what I was thinking when suggesting to scan the issues list for bug fixes when we're doing bugfix releases. Fixing bugs is less glamorous than implementing new features so I thought they might need some encouragement to be prioritized.

I'm really hesitant to adopt a workflow that requires multiple 'live' branches that need to be maintained as I don't think we have enough people to 1) robustly test the development branch and 2) deal with all of the git stuff that is necessary to keep thing organized. As always, I can be swayed if there are good arguments for it, but it feels to me like something like git flow is going to slow things down more than speed them up. It also feels like we're not really solving any of the issues, just pushing them somewhere less visible... we still have to deal with integrating everything, just now we have it all in this git branch rather than in PR queue. I feel like if we adhere more to using good PR labels and whatnot, there isn't much difference from a PR that's ready and merged into the `development` branch and a PR that's marked as 'reviewed' in the PR queue.

**Comment by jgosmann** [comment]
> The only bugs that would fall into this category for me are a) regressions, and b) bugs in things in master that have not been released yet.

There are some bugs that are better fixed earlier than later. For example #1048: The longer that bug persits, the more likely it is that some user ends up with a stale lock file that needs to manually removed. That is not obvious at all and the best for the user is to fix this in a timely manner so that no manual intervention on their side is necessary.

Also "bugs in things in master that have not been released yet" it might be good to fix these before a release before people start depending on the bug, thus making a fix more complicated. This won't apply to all bugs, but can happen in some instances.

Also I'm not sold that new features should have priority over bug fixes. Not entirely sure that anyone is saying that here, but it seems if no one is writing a PR to fix a bug, it won't get fixed? (Not saying that any bug should be fixed, but some users might care about bugs being fixed, that no one of the main developers cares a lot about? But of course putting the issue into a milestone doesn't necessarily mean that it gets fixed faster, but might give some more visibility to the issue).


> I don't see why we wouldn't just do a release and then do another one e.g. a week later when the critical bug has been fixed.

If the release breaks backwards compatibility, we might want to do a bugfix release for the earlier version?

> there isn't much difference from a PR that's ready and merged into the development branch and a PR that's marked as 'reviewed' in the PR queue.

To me it seems quite a difference, because the integration of the PRs isn't tested. Right now we tend to merge most of the PRs close to the release and not a lot of testing happens of the merged state before the release. #1044 is an example that was caused by the interaction of two PRs. There would be other ways to test this integration, like merging them to master and wait for a while before doing the release (assuming that most people in the lab use master), but there are probably some trade-offs. Having the PRs merged a while before the release makes it also easier for backends to test whether they are compatible with the upcoming version (and fix issues closer to the release or even before).

Not saying that all of this will eliminate all problems, but it might make it less likely to release bugs.

I have definitely to think a lot more about all of this (it's complicated and many things to consider), but to me it is not obvious (yet?) that all of the things proposed here are good set of rules to manage releases.

**Comment by jgosmann** [comment]
Also some more thoughts: In my undergrad we learned this technique to classify features (and bugs) with the MoSCoW(?) system: Putting each item into one of these categories:
* Must: These things need to be in the release, no matter what.
* Should: These should be in the release, but we don't manage to, we can leave them out.
* Could: If we have the time/resource: We can add them to the release, but we are not aiming to have them included.
* Won't?: This won't be included for whatever reason (to much work, not beneficial to the project, etc)

Maybe that can help with organizing what goes into a release?

Moreover, I don't think the list of features/bugfixes for a release needs to be fixed. We can have an initial list, but than renegotiate it at a later point to account for things that take longer or add other stuff that was finished earlier than expected. Though, this might not make much sense if only finished PRs are added to the list anyways.

**Comment by tbekolay** [comment]
> Moreover, I don't think the list of features/bugfixes for a release needs to be fixed.

I think this is the one thing I can't be swayed about. A constantly changing list is the main thing that makes releases drag on and on and makes my job way harder. With the exception of bugfixes, it's super frustrating to try to keep track of a changing list of things to do for a release. I'm always trying to keep in mind how far out we are from the next release and plan things around that, so putting something in the release pushes it out and means I'm constantly having to replan. Once we open the gate and say that a release list can change, what might be a small change for one developer gets multiplied by the number of developers that might want things in a release. That's was why my original proposal took a hardline about only having PRs in the list. If there aren't any PRs, then we wait until there are a few before making a release plan.

**Comment by jgosmann** [comment]
> so putting something in the release pushes it out and means I'm constantly having to replan

What if we have a fixed date for the release (or feature freeze)? So nothing can be added if it pushes the release back (except maybe critical bug fixes) and things that aren't finished in time will be dropped from the release and pushed back to the next release.

Would that allow you to plan? Or is the actual date of the release less important to you and the constancy of what will be included is what matters?


Maybe it helps to think of all of this as a few separate questions:

1. How do we prioritize what to work on?
2. How do we decide what will be in a release?
3. How do we decide when to do a release?

**Comment by jgosmann** [comment]
Ok, here's my proposal:

1. Bugs are classified in (at least) two categories: critical and non-critical.
   1. Bugs are non-critical by default.
   2. To mark a bug as critical someone (probably the reporter, but could be someone else) has to provide an argument why it is a critical bug. Some developer (or maintainer?) has to agree.
   3. Reasons for a critical bug could for example be (we can discuss what exactly should be included in this list and whether we need an exhaustive list):
      * A bugfix release breaks backwards compatibility; example: #1112
      * A timely fix is beneficial because "damage" accumulates; example: #1048
   4. Critical bugs should be rare. If they are not, our development model might need improvement or our criteria for critical bugs are too lax.
2. Issues for bugs can be assigned to milestones and block the release until resolved (I could be convinced to only allow this for critical bugs).
3. Issues for features must not be assigned to milestones.

Normal process after a new feature release (minor release) assuming no critical bugs are found:

1. For a time period yet to be specified no new release will be planned.
   * This allocates some time to get feedback and bug reports from users.
   * It also prevents feature releases being very close together (not sure if that's a concern).
2. After that time period we meet and evaluate:
   * Are there any bugs that should be fixed in a bug fix release? If so, the next release will be a bugfix release and the corresponding issues/PRs will be assigned to the milestone. Issues without a PR that require a significant amount of work should not be included.
   * Are there no bugs that warrant a bug fix release? If so, the next release will be a feature release. Only existing PRs ready for review and reviewed PRs can be selected and assigned to the milestone.
   * No bugs and no PRs? We wait until there are either bugs warranting a bug fix release or PRs warranting a feature release.
3. Reviewed PRs for the next release get continuously merged into master.
4. After the last PR for the milestone has been merged, we wait for a time period yet to be specified to see if any bugs caused by the interaction of PRs crop up and fix those if critical. (We could release this version as a release candidate, but it might not be worth the effort.)
5. We do the release and go back to step 1.

We could only use one of the time periods in step 1 or step 4 if those slow things down too much, but step 4 gives us opportunity to catch critical bugs before the release happens, whereas step 1 gives us the opportunity to collect feedback from users outside of the lab.

Given that a critical bug is discovered, this will trigger a bug fix release. If no next release is planned yet, the next release will be a bug fix release. If another release is already planned, we branch off of the last release, do the fix and release that as a bug fix release. We also cherry-pick the bug fix into the current master (so it will be included in the next planned release).


My intentions with this proposal are:
* Allow for timely fixes of critical bugs.
* Slight bias towards fixing bugs because they can be assigned to milestones, but only when decided so in a meeting planning for the next release 
* No single person can declare a bug as important enough to hold up release.
* Don't hold back releases with unfinished feature PRs.
* Have some integration testing.

I tried to keep this proposal as close as possible to @tbekolay's proposal. But I think it could easily be adopted to some of the proposals of @drasmuss and @hunse. For example, if there is no queue of PRs, we could continuously merge new PRs and decide at a meeting when there is enough stuff to do a new release.  Or instead of merging everything into master, we could have a develop branch.

**Comment by tbekolay** [comment]
Github just released a new feature, which is to [reorder issues/PRs in a milestone](https://github.com/blog/2206-reorder-issues-within-a-milestone). If we're thinking about prioritizing bugs, how about a "Bugs" milestone for bug issues that we order by priority?

**Comment by tbekolay** [comment]
We discussed this at the dev meeting last Friday; a summary of the decisions:

- Regarding when to do releases and whether they should be explicitly planned, we agreed to go with the retroactive approach to releases that Dan discussed [above](https://github.com/nengo/nengo/issues/1114#issuecomment-229181642). In short, we will review and merge pull requests as they roll in, and when master has diverged enough, we'll make a release. The type of release (patch, minor, or major) depends solely on the PRs merged (i.e., if we merge a non-breaking API change, the next version will up the minor number; if we merge a breaking API change, the next version will up the major number).
  - How do we determine when master has diverged "enough"? In general, by consensus... we'll look at what's been added to master since the last release every second lab meeting or so.
  - As a way to make releases more predictable, we should choose a maximum length of time between releases. The idea is that even if we've only merged minor improvements in that length of time, maybe someone is relying on those changes. We agreed to start with 1 month as the max length, but can adjust as we get a feel for how much happens in a month, and how much the releases are affecting people.

- Since the plan is to not plan, we could not come up with any good uses of milestones and so will no longer use milestones for anything. All of the uses that we could come up with (tracking high priority bugs, issues that are closed but might be resurrected in some form) are more effectively done with labels.

- Regarding tracking bugs of different priorities, we all agreed that we have been quite good, in general, at fixing important bugs that are reported (we didn't get any serious bug reports to core Nengo at the summer school, to my knowledge) so we're not going to add any more processes to triage and deal with bugs.

- One thing we have not been particularly good at is keeping in sync with backends after releases. We've talked about having a document for backend developers, which we really need to write. In that document, we should formalize the public API of the reference builder (for backends that use all of it, like OCL, and those that use parts of it, like SpiNNaker). We should consider assigning a version number to the backend API specifically as well, since that will make it explicit when backends will need updating.
  - These things are most salient for Nengo OCL, since it's the closest to the reference simulator, and shares the builder. For that reason, I will add a step to our release process to test the version being released with Nengo OCL. If it fails in a way that can be fixed in Nengo, then we should fix it in Nengo before releasing; if it can't, then we'll proceed with the release and make an issue on Nengo OCL to describe what changes are needed.
  - With the above new step, it should be possible for Nengo OCL to change its current policy of whitelisting the highest Nengo version that it works for to instead to assume that future Nengo versions will work properly. If we know of past versions of Nengo that don't work anymore, then those should be explicitly blacklisted.

With these decisions made, I think we can close this issue for now. If these processes end up not being great, then we can revisit them.

The remaining work is to document this in our developer docs, so I've added the 'document this' label to this issue. The 'document this' label is now being used in the same way as 'revise and resubmit', in the sense that the label is 'live' even if the issue is closed. Like this issue is now!
